test/enforce-schema.test.js

```diff
@@ -32,19 +32,37 @@ test('Should pass if no options passed and schema validation explicitly disabled
 
   await fastify.register(enforceSchema)
   
-  fastify.post('/foo', { schema: false }, (req, reply) => {
+  fastify.get('/foo', { schema: false }, (req, reply) => {
     reply.code(201).send('ok')
   })
 
   const res = await fastify.inject({
-    method: 'POST',
+    method: 'GET',
     url: '/foo'
   })
 
   t.equal(res.statusCode, 201)
   t.equal(res.payload, 'ok')
 })
 
+test('Should fail if no options passed: every POST should have a body unless explicitly set to false', async (t) => {
```

Comment:
You are absolutely right, body and response are enforced by default, so one would need to do the opposite, **explicitly** set them to false.

Reply:
Should the `params` schema also be required by default on routes with params?

If this is the case, I don't quite understand the necessity for the `required` option, since to my understanding the default behaviour is to require all three of `schema`, `body`, and `params`. Am I missing something or should we remove this option?
If we can remove it, I propose to replace it with a new `disabled` option to disable specific schemas from being validated.

Reply:
* **body** required for PATCH, PUT and POST
* params required (as you said), on routes with params
* response required on everything

This PR is/was suppose to introduce an ability for users to *explicitly* disable any of the 3 types. So in simple words, things are enforced by default, and e.g. if you don't respond with anything in some endpoint, you also have to explicitly state that.

Reply:
So I think you are more or less right. On the schemas that I previously mentioned, you would want an ability to primarily disable some of the schemas.

Reply:
Another question: in `utils.js` there is an export called `initialExcludes` which excludes a bunch of docs subdirectories by default. This behaviour isn't documented everywhere, so I assume this was primarily for the project you originally built this for. Should I:
1. Document this behaviour in the README
2. Remove these excludes and give you the code required to exclude them in the required project(s) so you can upgrade to the new version

Reply:
Great question. Those excludes are specific to [Fastify Swagger](https://github.com/fastify/fastify-swagger) plugin. It excludes the routes created by Swagger, and my sentiment was that such side effect should be handled in here, rather then a consumer of this library needing to manually exclude something that they even aren't owning themselves.

Reply:
So in simple words, I think we should leave those in there without needing to document it, as documenting internal library behavior that has no impact on the user would not introduce any clarity, and it would not be useful in any way.

Reply:
```
test('Should fail if body schema is missing', async (t) => {
  t.plan(1)

  const fastify = Fastify()

  await fastify.register(enforceSchema, { required: ['body'] })

  try {
    fastify.post('/foo', { schema: {} }, (req, reply) => {
      reply.code(201).send('ok')
    })
  } catch (error) {
    t.equal(error.message, 'POST: /foo is missing a body schema')
  }
})
```

If you are referring to this test, yes, this should fail, as you have provided a schema as an empty object literal.

Reply:
Will try to figure out what's causing the bug and post a commit by Monday.

Thanks so much for your elaboration, and Happy New Year to you and your family!

Reply:
> Will try to figure out what's causing the bug and post a commit by Monday.
> 
> Thanks so much for your elaboration, and Happy New Year to you and your family!

If anything, at the very least, I can thank you for putting effort at this level of detail into this. Whenever you are ready to merge, just let me know. Happy new year to you and your family as well. Looking forward to collaborate on various things in the upcoming year.

Cheers :) 🥳 🎈